clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -128,13 +127,6 @@ private void changeSubscription(Set<String> topicsToSubscribe) {
         if (!this.subscription.equals(topicsToSubscribe)) {
             this.subscription = topicsToSubscribe;
             this.groupSubscription.addAll(topicsToSubscribe);
-
-            // Remove any assigned partitions which are no longer subscribed to
-            for (Iterator<TopicPartition> it = assignment.keySet().iterator(); it.hasNext(); ) {
```

Comment:
I'm surprised that this didn't cause any unit test failures in `SubscriptionStateTest`. Haha, that probably shows that we have some gaps in our coverage. Maybe we could have test cases which validate that the assignment does not change with a call to `subscribe`, but does when we call `assignFromSubscribed`? Another one for regex would be nice as well. These cases are probably covered in `KafkaConsumerTest` already, but it's helpful to have unit testing as close as possible to the implementation.